src/index.ts

```diff
@@ -15,62 +25,115 @@ const buildRestClient = (apiClient: InfisicalApi, requestOptions?: RawAxiosReque
 	#authClient: AuthClient;
-	#basePath: string;
+	#tokenDetails?: TTokenDetails;
+	#authCredentials?: TAuthCredentials;
 
 	constructor(options?: InfisicalSDKOptions) {
-		this.#basePath = options?.siteUrl || "https://app.infisical.com";
+		this.sdkOptions = {
+			autoTokenRefresh: options?.autoTokenRefresh ?? true,
+			siteUrl: options?.siteUrl || "https://app.infisical.com"
+		};
 
 		this.#apiInstance = new InfisicalApi(
 			new Configuration({
-				basePath: this.#basePath
+				basePath: this.sdkOptions.siteUrl
 			})
 		);
 
-		this.#authClient = new AuthClient(this.authenticate.bind(this), this.#apiInstance, this.#basePath);
+		// We need to do bind(this) because the authenticate method is a private method, and usually you can't call private methods from outside the class.
+		this.#authClient = new AuthClient(this.#authenticator, this.#apiInstance, this.#tokenDetails?.accessToken);
 		this.#dynamicSecretsClient = new DynamicSecretsClient(this.#apiInstance, this.#requestOptions);
 		this.#secretsClient = new SecretsClient(this.#apiInstance, this.#requestOptions);
 		this.rest = () => buildRestClient(this.#apiInstance, this.#requestOptions);
+
+		if (this.sdkOptions.autoTokenRefresh) {
+			this.handleTokenLifecycle();
+		}
 	}
 
-	private authenticate(accessToken: string) {
-		this.#apiInstance = new InfisicalApi(
-			new Configuration({
-				basePath: this.#basePath,
-				accessToken
-			})
-		);
+	private handleTokenLifecycle = async () => {
+		const authMethodMap: Record<AuthMethod, () => Promise<InfisicalSDK>> = {
+			[AuthMethod.UniversalAuth]: async () =>
+				await this.#authClient.universalAuth.login(this.#authCredentials?.credentials as UniversalAuthCredentials),
+			[AuthMethod.AWSIam]: async () => await this.#authClient.awsIamAuth.login(this.#authCredentials?.credentials as AWSIamCredentials),
+			[AuthMethod.AccessToken]: async () =>
+				await this.#authClient.accessToken((this.#authCredentials?.credentials as AccessTokenCredentials).accessToken)
+		};
+
+		while (true) {
+			try {
+				await new Promise(resolve => setTimeout(resolve, 10_000));
+				if (!this.#tokenDetails || !this.#authCredentials) {
+					continue;
+				}
+
+				const now = new Date();
+				const timeSinceFetchInSeconds = (now.getTime() - this.#tokenDetails.fetchedTime.getTime()) / 1_000;
+				const firstFetchTimeInSeconds = (now.getTime() - this.#tokenDetails.firstFetchTime.getTime()) / 1_000;
 
-		this.#requestOptions = {
-			headers: {
-				Authorization: `Bearer ${accessToken}`
+				// If there's less than 15 seconds until the token expires, we reauthenticate early
+				if (firstFetchTimeInSeconds >= this.#tokenDetails.accessTokenMaxTTL - 15) {
+					await authMethodMap[this.#authCredentials.type]();
+					this.#tokenDetails.firstFetchTime = new Date();
+					// If there's less than 10 seconds until the token expires, we renew it early
+				} else if (timeSinceFetchInSeconds >= this.#tokenDetails.expiresIn - 10) {
```

Comment:
It might make more sense to see if we can renew the token first before we resort to the last hope which is re login

Reply:
This makes sense though. If the max TTL has been reached, there's no point trying to renew the token because it will always fail if max TTL has been reached